keps/sig-node/4603-tune-crashloopbackoff/README.md

```diff
@@ -0,0 +1,1222 @@
+canonical tracking issue has been open, some of which imply high levels of
+sophistication for kubelet to make different decisions based on system state,
+workload specific factors, or by detecting anomalous workload behavior. While
+this KEP does not rule out those directions in the future, the proposal herein
+focuses on the simpler, easily modeled changes that are designed to address the
+most common issues observed today. It is the desire of the author that the
+observability and benchmarking improvements instrumented as part of this
+proposal can serve as a basis for pursuing those types of improvements in the
+future, including as signals for end users to represent such solutions as
+necessary. Some analysis of these alternatives are included in the Alternatives
+Considered section below for future reference.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+* Improve pod restart backoff logic to better match the actual load it creates
+  and meet emerging use cases
+* Quantify and expose risks to node stability resulting from decreased and/or
+  hetereogeneous backoff configurations
+* Empirically derive defaults and allowable thresholds that approximate current
+  node stability
+* Provide a simple UX that does not require changes for the majority of
+  workloads
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* This effort is NOT intending to support fully user-specified configuration, to
+  cap risk to node stability
+* This effort is purposefully NOT implementing more complex heuristics by
+  kubelet (e.g. system state, workload specific factors, or by detecting
+anomalous workload behavior) to focus on better benchmarking and observability
+  and address common use cases with easily modelled changes first
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This design seeks to incorporate a three-pronged approach:
+
+1. Change the existing default backoff curve to stack more retries earlier, but
+   target the same amount of overall retries as the current behavior. This means
+   more restarts in the first 1 min or so, but later retries would be spaced out
+   further than they are today.
+2. Allow fast, flat-rate (0-10s + jitter) restarts when the exit code is 0, if
+   `restartPolicy: Always`.
+3. Provide a `restartPolicy: Rapid` option to configure even faster restarts for
+specific Pod/Container (particularly sidecar containers), regardless of exit
+code, reducing the max wait to 1 minute
```

Comment:
Notes from IRL convo, will integrate into KEP before resolving:

* We had a different starting opinion on how strongly we should interpret end user opt in to `Rapid`
* From the canonical issue, 1 minute is commonly referred to as the max wait that is acceptable
* I'm ok with giving `Rapid` and "Succeeded" containers the same initial value to make their restart distributions simpler to follow, but failing `Rapid` containers should still backoff (because they could be legitimately crashing)
* We'll really be finalizing the initial and max cap after benchmarking, and while I was starting more near initial=250ms / cap=1 minute, your had thoughts more like initial="as fast as kubelet can handle maybe 0" / cap=10-30 s, so benchmarks should at least represent those cases so we can evaluate between them

Reply:
Updated in https://github.com/kubernetes/enhancements/pull/4604/commits/385894f19f107994a94aa1825c171e19d948dae5